core/parachain/tasks_sequence.hpp

```diff
@@ -0,0 +1,144 @@
+    template <typename F>
+    [[maybe_unused]] void operator()(F &&func);
+  };
+
+  template <>
+  struct ThreadQueueContext<std::weak_ptr<boost::asio::io_context>> {
+    using Type = std::weak_ptr<boost::asio::io_context>;
+    Type t;
+
+    template <typename D>
+    ThreadQueueContext(D &&arg) : t{std::forward<D>(arg)} {}
+
+    template <typename F>
+    void operator()(F &&func) {
+      if (auto call_context = t.lock()) {
+        boost::asio::post(*call_context, std::forward<F>(func));
+      }
+    }
+  };
+
+  template <>
+  struct ThreadQueueContext<std::shared_ptr<boost::asio::io_context>>
+      : ThreadQueueContext<std::weak_ptr<boost::asio::io_context>> {
+    template <typename D>
+    ThreadQueueContext(D &&arg)
+        : ThreadQueueContext<std::weak_ptr<boost::asio::io_context>>(
+            std::forward<D>(arg)) {}
+  };
+
+  template <typename T>
+  auto createThreadQueueContext(T &&t) {
+    return ThreadQueueContext<std::decay_t<T>>{std::forward<T>(t)};
+  }
+
+  template <typename C, typename F>
+  auto createTask(C &&c, F &&f) {
+    return std::make_pair(createThreadQueueContext(std::forward<C>(c)),
+                          std::forward<F>(f));
+  }
+
+  template <typename T, typename K, typename... Args>
+  void sequence(std::pair<T, K> &&t, Args &&...args) {
+    __internal_contextCall(std::move(t.first),
+                           [func{std::move(t.second)},
+                            forwarding_func{__internal_bindArgs(
+                                std::forward<Args>(args)...)}]() mutable {
+                             forwarding_func(func());
+                           });
+  }
+
+  template <typename T, typename K>
+  void sequence(std::pair<T, K> &&t) {
+    __internal_contextCall(std::move(t.first), std::move(t.second));
+  }
+
+  template <typename T, typename F>
+  void __internal_contextCall(ThreadQueueContext<T> &&t, F &&f) {
+    t(std::forward<F>(f));
+  }
+
+  template <typename R, typename T, typename K>
+  void __internal_executeI(R &&r, std::pair<T, K> &&t) {
+    if (std::forward<R>(r).has_value()) {
+      __internal_contextCall(
+          std::move(t.first),
+          [r{std::forward<R>(r).value()}, func{std::move(t.second)}]() mutable {
+            func(std::move(r));
+          });
+    }
+  }
+
+  template <typename R, typename T, typename K, typename... Args>
+  void __internal_executeI(R &&r, std::pair<T, K> &&t, Args &&...args) {
+    if (std::forward<R>(r).has_value()) {
+      __internal_contextCall(std::move(t.first),
+                             [r{std::forward<R>(r).value()},
+                              func{std::move(t.second)},
+                              forwarding_func{__internal_bindArgs(
+                                  std::forward<Args>(args)...)}]() mutable {
+                               forwarding_func(func(std::move(r)));
+                             });
+    }
+  }
+
+  template <typename... Args>
+  auto __internal_bindArgs(Args &&...args) {
```

Comment:
Perhaps would be better to move entrails `__internal_*` to namespase `kagome::thread::details`